config.go

```diff
@@ -6,53 +6,130 @@ import (
 	"net/http"
 	"net/url"
 	"os"
+	"reflect"
 )
 
 type Opt interface {
 	apply(*cfg)
+	Type() string
 }
 
 type cfg struct {
 	controllerUrl  string
 	authToken      string
+	authType       string
 	httpAuthWriter httpAuthWriter
 	logger         *slog.Logger
 }
 
 type clientOpt struct{ fn func(*cfg) }
 
+// controllerUrlOpt is an option to set the controller url for the client
+type controllerUrlOpt struct {
+	controllerUrl string
+}
+
+func (o *controllerUrlOpt) apply(c *cfg) {
+	c.controllerUrl = o.controllerUrl
+}
+
+func (o *controllerUrlOpt) Type() string {
+	return "controllerUrl"
+}
+
+// authTokenOpt is an option to set the auth token for the client
+type authTokenOpt struct {
+	authToken string
+}
+
+func (o *authTokenOpt) apply(c *cfg) {
+	c.authToken = o.authToken
+}
+
+func (o *authTokenOpt) Type() string {
+	return "authToken"
+}
+
+// authTypeOpt is an option to set the auth type for the client
+type authTypeOpt struct {
+	authType string
+}
+
+func (o *authTypeOpt) apply(c *cfg) {
+	c.authType = o.authType
+}
+
+func (o *authTypeOpt) Type() string {
+	return "authType"
```

Comment:
these strings should be named something like `const AuthType = "authType"`

its reference by string again on line `108`

Reply:
new commit